### PR TITLE
mpi/util: do not use non-gpu safe memcpy in MPIR_Localcopy

### PR DESCRIPTION
**Pull request by gcongiu**
## Pull Request Description
Remove any use of non GPU safe memory copy operations, i.e., direct use
of `MPIR_Memcpy`. Instead convert the existing code to only use
`MPIR_Typerep_pack/unpack` which are optimized for GPU and contiguous data
movement as well.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by gcongiu** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by gcongiu** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: src/mpi/misc/utils.c [hunk 1]
```diff
@@ -15,6 +15,8 @@ int MPIR_Localcopy(const void *sendbuf, MPI_Aint sendcount, MPI_Datatype sendtyp
     int sendtype_iscontig, recvtype_iscontig;
     MPI_Aint sendsize, recvsize, sdata_sz, rdata_sz, copy_sz;
     MPI_Aint true_extent, sendtype_true_lb, recvtype_true_lb;
+    char *buf = NULL;
+    MPL_pointer_type_t send_type, recv_type;
```
Comment (pavanbalaji) [inline src/mpi/misc/utils.c:19]: These variables can sit in the noncontig else branch.

File: src/mpi/misc/utils.c [hunk 2]
```diff
@@ -26,6 +28,8 @@ int MPIR_Localcopy(const void *sendbuf, MPI_Aint sendcount, MPI_Datatype sendtyp
     sdata_sz = sendsize * sendcount;
     rdata_sz = recvsize * recvcount;
 
+    send_type = recv_type = MPL_GPU_POINTER_UNREGISTERED_HOST;
```
Comment (pavanbalaji) [inline src/mpi/misc/utils.c:31]: Why is this initialization needed?

File: src/mpi/misc/utils.c [hunk 3]
```diff
@@ -40,43 +44,38 @@ int MPIR_Localcopy(const void *sendbuf, MPI_Aint sendcount, MPI_Datatype sendtyp
...
-                                  ((char *) sendbuf + sendtype_true_lb), copy_sz);
-#endif
-        MPIR_Memcpy(((char *) recvbuf + recvtype_true_lb),
-                    ((char *) sendbuf + sendtype_true_lb), copy_sz);
+    if (recvtype_iscontig) {
+        MPI_Aint actual_pack_bytes;
+        MPIR_Typerep_pack(sendbuf, sendcount, sendtype, 0, (char *) recvbuf + recvtype_true_lb,
+                          copy_sz, &actual_pack_bytes);
+        MPIR_ERR_CHKANDJUMP(actual_pack_bytes != copy_sz, mpi_errno, MPI_ERR_TYPE,
+                            "**dtypemismatch");
```
Comment (pavanbalaji) [inline src/mpi/misc/utils.c:58]: Why is this branch moved up?  It's an unnecessary change.

File: src/mpi/misc/utils.c [hunk 4]
```diff
@@ -40,43 +44,38 @@ int MPIR_Localcopy(const void *sendbuf, MPI_Aint sendcount, MPI_Datatype sendtyp
...
-        char *buf;
         intptr_t sfirst;
         intptr_t rfirst;
 
-        MPIR_CHKLMEM_MALLOC(buf, char *, COPY_BUFFER_SZ, mpi_errno, "buf", MPL_MEM_BUFFER);
+        MPL_gpu_query_pointer_type(sendbuf, &send_type);
+        MPL_gpu_query_pointer_type(recvbuf, &recv_type);
+
+        if (send_type == MPL_GPU_POINTER_DEV && recv_type == MPL_GPU_POINTER_DEV) {
+            MPL_gpu_malloc((void **) &buf, COPY_BUFFER_SZ);
```
Comment (pavanbalaji) [inline src/mpi/misc/utils.c:73]: Doesn't GPU malloc require an actual device ID?

File: src/mpi/misc/utils.c [hunk 5]
```diff
@@ -40,43 +44,38 @@ int MPIR_Localcopy(const void *sendbuf, MPI_Aint sendcount, MPI_Datatype sendtyp
...
+        } else if (send_type == MPL_GPU_POINTER_DEV || recv_type == MPL_GPU_POINTER_DEV) {
+            MPL_gpu_malloc_host((void **) &buf, COPY_BUFFER_SZ);
```
Comment (pavanbalaji) [inline src/mpi/misc/utils.c:75]: This doesn't look right.  We only need the host buffer in the case where the send type is noncontiguous and is sitting on the host.

File: src/mpi/misc/utils.c (hunk 4 above)
Comment (gcongiu) [reply src/mpi/misc/utils.c:73]: Nvidia and AMD do not need one but I forgot that Intel does

File: src/mpi/misc/utils.c (hunk 5 above)
Comment (gcongiu) [reply src/mpi/misc/utils.c:75]: This branch is for non-contiguous send and recv types. If either one of the two is in GPU memory we need to allocate host registered memory. Otherwise pack (unpack) will do additional copy to (from) OS bounce buffer. Am I recollecting correctly?

File: src/mpi/misc/utils.c (hunk 3 above)
Comment (gcongiu) [reply src/mpi/misc/utils.c:58]: Right

File: src/mpi/misc/utils.c (hunk 2 above)
Comment (gcongiu) [reply src/mpi/misc/utils.c:31]: We check for memory type when freeing the pack/unpack buffer at the end. This is to initialize the default case, i.e., pageable host memory

File: src/mpi/misc/utils.c (hunk 4 above)
Comment (pavanbalaji) [reply src/mpi/misc/utils.c:73]: No, CUDA and HIP need one too.

File: src/mpi/misc/utils.c (hunk 2 above)
Comment (pavanbalaji) [reply src/mpi/misc/utils.c:31]: Why is there a default case?  The query will return the right type, even if that's unregistered host.

File: src/mpi/misc/utils.c (hunk 5 above)
Comment (pavanbalaji) [reply src/mpi/misc/utils.c:75]: Hmm.  Maybe you are right.  I miscalculated.  We probably do need both a GPU buffer and host buffer.

File: src/mpi/misc/utils.c (hunk 2 above)
Comment (gcongiu) [reply src/mpi/misc/utils.c:31]: Right but we need to allocate memory only in the case in which both types are non contiguous, there is no need to query memory for other cases. So I would just set this manually to unregister host and query only if needed.

Comment (pavanbalaji) [reply src/mpi/misc/utils.c:31]: But you are not using them in any other case, right?  They are only used in the noncontig case.

Comment (gcongiu) [reply src/mpi/misc/utils.c:31]: If pack/unpack fails I still need to free the allocated buffer when exiting the function, so I need to know what buffer was allocated depending on the send/recv memory type.

Comment (pavanbalaji) [reply src/mpi/misc/utils.c:31]: If the pack/unpack in the contig cases fail, there's no buffer to free.

Comment (gcongiu) [reply src/mpi/misc/utils.c:31]: Yes. There are two options: keep everything local to the branch and in case of type mismatch clear the `buf` before `MPIR_ERR_CHKANDJUMP` or (as I do now) clear the buffer in `fn_fail`. The second approach is what we typically do.

Comment (pavanbalaji) [reply src/mpi/misc/utils.c:31]: Alright, I think we spent more time arguing about it than what it deserves.  :-)  I'll ignore it.

File: src/mpi/misc/utils.c [hunk 6]
```diff
@@ -72,11 +63,21 @@ int MPIR_Localcopy(const void *sendbuf, MPI_Aint sendcount, MPI_Datatype sendtyp
...
         intptr_t sfirst;
         intptr_t rfirst;
 
-        MPIR_CHKLMEM_MALLOC(buf, char *, COPY_BUFFER_SZ, mpi_errno, "buf", MPL_MEM_BUFFER);
+        MPL_gpu_query_pointer_type(sendbuf, &send_type);
+        MPL_gpu_query_pointer_type(recvbuf, &recv_type);
+
+        if (send_type == MPL_GPU_POINTER_DEV && recv_type == MPL_GPU_POINTER_DEV) {
+            MPL_gpu_dev_id_t devid;
+            MPL_gpu_get_device_id(&devid);
```
Comment (pavanbalaji) [inline src/mpi/misc/utils.c:74]: What is this function?  Where is this getting a device ID from?

Comment (gcongiu) [reply src/mpi/misc/utils.c:74]: Hmmm, right now it is considering the current GPU context. So in principle I see two cases: `1` - the `MPIR_Localcopy` is local, meaning we are moving data within the process address space (either D2D, D2H or H2D); `2` - the `MPIR_Localcopy` is not local, meaning we are moving data from another process address space (on a different GPU context) into the process address space (possibly to the GPU memory). In this case I guess allocating the temporary buffer in the target process GPU should be fine. The device id query function should probably be more generic still, also taking a buffer as input and returning the device id to which that buffer belongs to. Is this your concern here?

Comment (pavanbalaji) [reply src/mpi/misc/utils.c:74]: What is "current GPU context"?  There is no such thing for L0 API.

Comment (pavanbalaji) [reply src/mpi/misc/utils.c:74]: Besides, even for CUDA, "current GPU context" is irrelevant here.  The only thing that matters is which device my buffers map to.

Comment (gcongiu) [reply src/mpi/misc/utils.c:74]: You are right. I will fix this.

**Comment by gcongiu** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by gcongiu** [comment]
test:mpich/ch3/most

**Comment by gcongiu** [comment]
All test passed:
https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch4-ucx/887/
https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch4-ofi/1149/
https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch3-tcp/598/
https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch3-sock/418/

Will rebase on master and merge.
